src/parcsr_ls/par_interp_device.c

```diff
@@ -1009,4 +1012,497 @@ hypre_BoomerAMGBuildDirInterp_getcoef_v2( HYPRE_Int   nr_of_rows,
+                      thrust::make_counting_iterator(num_cols_A_offd),
+                      mark_P_offd_idx,
+                      offd_map_P_to_A,
+                      equal<HYPRE_Int>(1) );
+   hypre_TFree(mark_P_offd_idx, HYPRE_MEMORY_DEVICE);
+
+   /* also get an inverse mapping from A offd indices to P offd indices */ 
+   /* offd_map_A_to_P[ A offd idx ] = -1 if not a P idx, else P offd idx */
+   HYPRE_Int *offd_map_A_to_P = hypre_TAlloc(HYPRE_Int, num_cols_A_offd, HYPRE_MEMORY_DEVICE);
+   HYPRE_THRUST_CALL( fill_n,
+                      offd_map_A_to_P,
+                      num_cols_A_offd,
+                      -1 );
+   HYPRE_THRUST_CALL( scatter,
+                      thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(num_cols_P_offd),
+                      offd_map_P_to_A,
+                      offd_map_A_to_P );
+
+   /* use inverse mapping above to map P_offd_j */
+   HYPRE_THRUST_CALL( gather,
+                      P_offd_j_temp_compressed,
+                      P_offd_j_temp_compressed + nnz_offd,
+                      offd_map_A_to_P,
+                      P_offd_j );
+   hypre_TFree(P_offd_j_temp_compressed, HYPRE_MEMORY_DEVICE);
+   hypre_TFree(offd_map_A_to_P, HYPRE_MEMORY_DEVICE);
+
+   /* setup col_map_offd for P */
+   col_map_offd_P_device = hypre_CTAlloc(HYPRE_BigInt, num_cols_P_offd, HYPRE_MEMORY_DEVICE);
+   col_map_offd_P = hypre_CTAlloc(HYPRE_BigInt, num_cols_P_offd, HYPRE_MEMORY_HOST);
+   HYPRE_THRUST_CALL( gather,
+                      offd_map_P_to_A,
+                      offd_map_P_to_A + num_cols_P_offd,
+                      fine_to_coarse_offd,
+                      col_map_offd_P_device);
+   hypre_TMemcpy(col_map_offd_P, col_map_offd_P_device, HYPRE_Int, num_cols_P_offd, HYPRE_MEMORY_HOST, HYPRE_MEMORY_DEVICE);
+   hypre_TFree(offd_map_P_to_A, HYPRE_MEMORY_DEVICE);
+   hypre_TFree(col_map_offd_P_device, HYPRE_MEMORY_DEVICE);
+
+   /* Now, we should have everything of Parcsr matrix P */
+   P = hypre_ParCSRMatrixCreate(comm,
+                                hypre_ParCSRMatrixGlobalNumCols(A), /* global num of rows */
+                                total_global_cpts, /* global num of cols */
+                                hypre_ParCSRMatrixColStarts(A), /* row_starts */
+                                num_cpts_global, /* col_starts */
+                                num_cols_P_offd, /* num cols offd */
+                                nnz_diag,
+                                nnz_offd);
+
+   P_diag = hypre_ParCSRMatrixDiag(P);
+   hypre_CSRMatrixData(P_diag) = P_diag_data;
+   hypre_CSRMatrixI(P_diag)    = P_diag_i;
+   hypre_CSRMatrixJ(P_diag)    = P_diag_j;
+
+   P_offd = hypre_ParCSRMatrixOffd(P);
+   hypre_CSRMatrixData(P_offd) = P_offd_data;
+   hypre_CSRMatrixI(P_offd)    = P_offd_i;
+   hypre_CSRMatrixJ(P_offd)    = P_offd_j;
+   /* P does not own ColStarts, since A does */
+   hypre_ParCSRMatrixOwnsRowStarts(P) = 0;
```

Comment:
`hypre_ParCSRMatrixOwnsRowStarts` is removed in #392

Reply:
Yes, I caught this when I merged the latest master again. Thanks.